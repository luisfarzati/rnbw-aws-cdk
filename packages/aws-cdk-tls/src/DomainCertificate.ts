import * as acm from "@aws-cdk/aws-certificatemanager";
import * as route53 from "@aws-cdk/aws-route53";
import * as cdk from "@aws-cdk/core";

export type DomainCertificateProps = {
  /**
   * Domain name of this certificate.
   */
  domain: string;
  /**
   * Domain name associated to the Hosted Zone where the DNS validation record
   * will be created. Optional; defaults to the given `domain` value.
   * Not used if `hostedZone` is specified.
   */
  zoneDomain?: string;
  /**
   * An instance of IHostedZone; overrides `zoneDomain`.
   */
  hostedZone?: route53.IHostedZone;
  /**
   * Region where this certificate will be created. Certificates used for
   * CloudFront distributions require the region to be `us-east-1`.
   */
  certificateRegion?: string;
};

/**
 * Represents a TLS certificate for a given domain.
 */
export class DomainCertificate extends cdk.Construct {
  private readonly _certificate: acm.ICertificate;

  constructor(
    parent: cdk.Construct,
    id: string,
    props: DomainCertificateProps
  ) {
    super(parent, id);

    const { domain, zoneDomain, hostedZone, certificateRegion } = props;

    //
    // Route53 HostedZone
    //
    const zone =
      hostedZone ||
      route53.HostedZone.fromLookup(this, "DomainCertificateHostedZone", {
        domainName: zoneDomain || domain
      });

    //
    // ACM Certificate
    //
    const certificate = new acm.DnsValidatedCertificate(
      this,
      "DomainCertificate",
      {
        domainName: domain,
        validationMethod: acm.ValidationMethod.DNS,
        hostedZone: zone,
        region: certificateRegion
      }
    );

    this._certificate = certificate;
  }

  get certificate() {
    return this._certificate;
  }
}
