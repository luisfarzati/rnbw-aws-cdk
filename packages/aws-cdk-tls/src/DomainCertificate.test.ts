import * as cdk from "@aws-cdk/core";
import { SynthUtils } from "@aws-cdk/assert";
import { DomainCertificate } from "./DomainCertificate";

const testEnv = {
  env: {
    region: "us-west-1",
    account: "test-account"
  }
};

const createStack = () => new cdk.Stack(undefined, undefined, testEnv);

test("Matches CloudFormation snapshot", () => {
  const stack = createStack();

  new DomainCertificate(stack, "DomainCertificateTest", {
    domain: "www.domain.com",
    zoneDomain: "domain.com",
    certificateRegion: "us-east-1"
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test("Has resource properties", () => {
  const stack = createStack();

  const domain = new DomainCertificate(stack, "DomainCertificateTest", {
    domain: "domain.com"
  });

  expect(domain.certificate).toBeDefined();
});
