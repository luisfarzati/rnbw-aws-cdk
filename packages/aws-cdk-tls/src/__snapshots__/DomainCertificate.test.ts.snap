// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches CloudFormation snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters0b9723d54b7db3fbfc1a143398b02392e5fe080a68535480782a949b4372d0f0ArtifactHashF8F836D1": Object {
      "Description": "Artifact hash for asset \\"0b9723d54b7db3fbfc1a143398b02392e5fe080a68535480782a949b4372d0f0\\"",
      "Type": "String",
    },
    "AssetParameters0b9723d54b7db3fbfc1a143398b02392e5fe080a68535480782a949b4372d0f0S3Bucket0997A4A0": Object {
      "Description": "S3 bucket for asset \\"0b9723d54b7db3fbfc1a143398b02392e5fe080a68535480782a949b4372d0f0\\"",
      "Type": "String",
    },
    "AssetParameters0b9723d54b7db3fbfc1a143398b02392e5fe080a68535480782a949b4372d0f0S3VersionKey83D9C166": Object {
      "Description": "S3 key for asset version \\"0b9723d54b7db3fbfc1a143398b02392e5fe080a68535480782a949b4372d0f0\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "DomainCertificateTestDomainCertificateCertificateRequestorFunctionD5719468": Object {
      "DependsOn": Array [
        "DomainCertificateTestDomainCertificateCertificateRequestorFunctionServiceRoleDefaultPolicyAA3ACB74",
        "DomainCertificateTestDomainCertificateCertificateRequestorFunctionServiceRoleED24A52E",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters0b9723d54b7db3fbfc1a143398b02392e5fe080a68535480782a949b4372d0f0S3Bucket0997A4A0",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0b9723d54b7db3fbfc1a143398b02392e5fe080a68535480782a949b4372d0f0S3VersionKey83D9C166",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0b9723d54b7db3fbfc1a143398b02392e5fe080a68535480782a949b4372d0f0S3VersionKey83D9C166",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.certificateRequestHandler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "DomainCertificateTestDomainCertificateCertificateRequestorFunctionServiceRoleED24A52E",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "DomainCertificateTestDomainCertificateCertificateRequestorFunctionServiceRoleDefaultPolicyAA3ACB74": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "acm:RequestCertificate",
                "acm:DescribeCertificate",
                "acm:DeleteCertificate",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:GetChange",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:changeResourceRecordSets",
              "Effect": "Allow",
              "Resource": "arn:aws:route53:::hostedzone/DUMMY",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DomainCertificateTestDomainCertificateCertificateRequestorFunctionServiceRoleDefaultPolicyAA3ACB74",
        "Roles": Array [
          Object {
            "Ref": "DomainCertificateTestDomainCertificateCertificateRequestorFunctionServiceRoleED24A52E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DomainCertificateTestDomainCertificateCertificateRequestorFunctionServiceRoleED24A52E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DomainCertificateTestDomainCertificateCertificateRequestorResource9D9F6E7A": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DomainName": "www.domain.com",
        "HostedZoneId": "DUMMY",
        "Region": "us-east-1",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "DomainCertificateTestDomainCertificateCertificateRequestorFunctionD5719468",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
